version: '3.6'

services:
  database:
    container_name: backend_db
    image: postgres:14
    restart: unless-stopped
    volumes:
      - track-and-trace:/var/lib/postgresql/data
    networks:
      backend-network:
        ipv4_address: 172.45.0.3
    ports:
      - 5432:5432
    environment:
      POSTGRES_USER: "user"
      POSTGRES_PASSWORD: "password"
      POSTGRES_DB: "db"
      PGDATA: /var/lib/postgresql/data/db/
  backend:
    container_name: backend_app
    build:
      context: ./backend
      dockerfile: docker/dev.Dockerfile
    depends_on:
      - database
    volumes:
      # Binds volume to enable working through docker locally
      - type: bind
        source: ./backend/src/
        target: /app/src/
      - type: bind
        source: ./backend/package.json
        target: /app/package.json
      - type: bind
        source: ./backend/tsconfig.json
        target: /app/tsconfig.json
      - type: bind
        source: ./backend/prisma/
        target: /app/prisma/
      - type: bind
        source: ./backend/services/
        target: /app/services/
    networks:
      backend-network:
        ipv4_address: 172.45.0.2
      frontend-network:
        ipv4_address: 172.46.0.2
    ports:
      - 5000:5000
    environment:
      NODE_ENV: 'local'
      PORT: 5000
      DATABASE_URL: 'postgresql://user:password@172.45.0.1:5432/db?schema=public'
  frontend:
    container_name: frontend_app
    build:
      context: ./frontend
      dockerfile: docker/dev.Dockerfile
    depends_on:
      - backend
    volumes:
      # Binds volume to enable working through docker locally
      - type: bind
        source: ./frontend
        target: /usr/src/app
    networks:
      frontend-network:
        ipv4_address: 172.46.0.3
    ports:
      - 3000:3000
    environment:
      NODE_ENV: 'local'
      API_URL: 'http://172.46.0.2:5000'
volumes:
  track-and-trace:
networks:
  backend-network:
    ipam:
      driver: default
      config:
        - subnet: 172.45.0.0/16
  frontend-network:
    ipam:
      driver: default
      config:
        - subnet: 172.46.0.0/16
